decimaldate.js

```diff
@@ -9,129 +9,157 @@ decimaldate.DECIMALPLACES = 6;
+    // months and day as 2 digits
+    // ISO 8601 shift year<=0 by 1, 0=1BCE, -1=2BCE
+    const monthstring = monthint.toString().padStart(2, '0');
+    const daystring = dayint.toString().padStart(2, '0');
+    let yearstring;
+    if (yearint > 0) yearstring = yearint.toString().padStart(4, '0');  // just the year as 4 digits
+    else if (yearint == -1) yearstring = (Math.abs(yearint + 1).toString().padStart(4, '0'));  // BCE offset by 1 but do not add a - sign
+    else yearstring = '-' + (Math.abs(yearint + 1).toString().padStart(4, '0'));  // BCE offset by 1 and add  - sign
+
+    return `${yearstring}-${monthstring}-${daystring}`;
 };
 
 
-decimaldate.isvalidmonth = (monthstring) => {
-    validmonths = ['01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12'];
-    return validmonths.indexOf(monthstring) != -1;
-};
+decimaldate.isvalidmonthday = (yearint, monthint, dayint) => {
+    if (yearint != parseInt(yearint) || yearint == 0) return false;
+    if (monthint != parseInt(monthint)) return false;
+    if (dayint != parseInt(dayint)) return false;
 
+    if (monthint < 1 || monthint > 12) return false;
+    if (dayint < 1) return false;
+
+    const dtm = decimaldate.daysinmonth(yearint, monthint);
+    if (! dtm) return false;
+    if (dayint > dtm) return false;
 
-decimaldate.isvalidmonthday = (yearstring, monthstring, daystring) => {
-    days = parseInt(daystring);
-    if (isNaN(days)) return false;
-    if (days < 0) return false;
-    const maxdays = decimaldate.daysinmonth(yearstring, monthstring);
-    if (! maxdays || days > maxdays) return false;
     return true;
 };
 
 
-decimaldate.proportionofdayspassed = (yearstring, monthstring, daystring) => {
-    // count the number of days to get to this day of this month
+decimaldate.proportionofdayspassed = (yearint, monthint, dayint) => {
+    // count the number of days to get through the prior months
     let dayspassed = 0;
-    ['01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12']
-    .forEach((tms) => {
-        if (tms < monthstring) dayspassed += decimaldate.daysinmonth(yearstring, tms);
-    });
-    dayspassed += parseInt(daystring);
-
-    // subtract 1 cuz day 0 is January 1 and not January 0
-    // add 0.5 to get us 12 noon
-    dayspassed -= 1;
-    dayspassed += 0.5;
-
-    // divide by days in year, to get decimal portion since noon of Jan 1
-    const dty = decimaldate.daysinyear(yearstring);
+    for (let m = 1; m < monthint; m++) {
+        const dtm = decimaldate.daysinmonth(yearint, m);
+        dayspassed += dtm;
+    }
+
+    // add the leftover days not in a prior month
+    // but minus 0.5 to get us to noon of the target day, as opposed to the end of the day
+    dayspassed = dayspassed + dayint - 0.5;
+
+    // divide by days in year, to get decimal portion
+    // even January 1 is 0.5 days in since we snap to 12 noon
+    const dty = decimaldate.daysinyear(yearint);
     return dayspassed / dty;
 };
```

Comment:
I don’t have a lot of context about why we calculate the proportion manually based on the number of days in the year. Wouldn’t it be more straightforward to generate two `Date` objects for January 1 of the current and next year and convert to milliseconds? I think this would be easier to reason about than something more manual. For example, we don’t have to care about leap years. We can still snap to noon if desired.

https://github.com/OpenHistoricalMap/ohm-embed-1yr/blob/e09dbeff35bb40e4fd38c5b8cec3c29551898bb3/embed-map-single-year.js#L105-L108